### PR TITLE
[Provisioner] Reduce parallelism for file mounts based on fd limit

### LINKED ISSUE #3002: [core] Launching a large cluster fails with `Too many open files`
Trying to launch a 75 node cluster fails with `Too many open files` during setup phase. Increasing ulimit with `ulimit -n 65535` does not help, stays stuck on `⠙ Preparing SkyPilot runtime (2/3 - dependencies)` for ~45 min+. 

Repro:

```
$ sky launch --num-nodes=75 --cpus 2+ --use-spot --down -c test --cloud aws

...
D 01-19 08:00:11 provisioner.py:503] Stacktrace:
D 01-19 08:00:11 provisioner.py:503] Traceback (most recent call last):
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/provisioner.py", line 496, in post_provision_runtime_setup
D 01-19 08:00:11 provisioner.py:503]     return _post_provision_setup(cloud_name,
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/provisioner.py", line 409, in _post_provision_setup
D 01-19 08:00:11 provisioner.py:503]     instance_setup.internal_file_mounts(cluster_name.name_on_cloud,
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/instance_setup.py", line 84, in wrapper
D 01-19 08:00:11 provisioner.py:503]     return func(*args, **kwargs)
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/instance_setup.py", line 403, in internal_file_mounts
D 01-19 08:00:11 provisioner.py:503]     _parallel_ssh_with_cache(
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/instance_setup.py", line 124, in _parallel_ssh_with_cache
D 01-19 08:00:11 provisioner.py:503]     return [future.result() for future in results]
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/instance_setup.py", line 124, in <listcomp>
D 01-19 08:00:11 provisioner.py:503]     return [future.result() for future in results]
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/tools/anaconda3/lib/python3.9/concurrent/futures/_base.py", line 439, in result
D 01-19 08:00:11 provisioner.py:503]     return self.__get_result()
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/tools/anaconda3/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
D 01-19 08:00:11 provisioner.py:503]     raise self._exception
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/tools/anaconda3/lib/python3.9/concurrent/futures/thread.py", line 58, in run
D 01-19 08:00:11 provisioner.py:503]     result = self.fn(*self.args, **self.kwargs)
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/metadata_utils.py", line 45, in wrapper
D 01-19 08:00:11 provisioner.py:503]     return function(*args, **kwargs)
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/instance_setup.py", line 401, in _setup_node
D 01-19 08:00:11 provisioner.py:503]     _internal_file_mounts(common_file_mounts, runner, log_path)
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/instance_setup.py", line 68, in retry
D 01-19 08:00:11 provisioner.py:503]     raise e
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/instance_setup.py", line 65, in retry
D 01-19 08:00:11 provisioner.py:503]     return func(*args, **kwargs)
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/instance_setup.py", line 372, in _internal_file_mounts
D 01-19 08:00:11 provisioner.py:503]     rc, stdout, stderr = runner.run(mkdir_command,
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/command_runner.py", line 343, in run
D 01-19 08:00:11 provisioner.py:503]     return log_lib.run_with_log(' '.join(command),
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/skylet/log_lib.py", line 188, in run_with_log
D 01-19 08:00:11 provisioner.py:503]     with subprocess.Popen(cmd,
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/tools/anaconda3/lib/python3.9/subprocess.py", line 951, in __init__
D 01-19 08:00:11 provisioner.py:503]     self._execute_child(args, executable, preexec_fn, close_fds,
D 01-19 08:00:11 provisioner.py:503]   File "/Users/romilb/tools/anaconda3/lib/python3.9/subprocess.py", line 1720, in _execute_child
D 01-19 08:00:11 provisioner.py:503]     errpipe_read, errpipe_write = os.pipe()
D 01-19 08:00:11 provisioner.py:503] OSError: [Errno 24] Too many open files
```

[Full logs](https://gist.github.com/romilbhardwaj/2f65215442030012a5d0a87e8d9b1b3a).

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #3002. This should be compatible with #3004.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch -c test-100-nodes --cloud aws --num-nodes 100 --cpus 2 echo hi`
  - [x] `sky down test-100-nodes`
  - [x] `sky status -r test-100-nodes` (tried again)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Comment by concretevitamin** [comment]
With the latest commit, 100 nodes successfully launched!
```
D 01-20 09:01:08 provisioner.py:124] Provisioning 'dbg' took 91.64 seconds.

D 01-20 09:01:12 provisioner.py:366] Waiting for SSH to be available for 'dbg' ...
D 01-20 09:01:25 provisioner.py:268] Waiting for SSH to 15.165.9.140. Try: ssh -T -i '~/.ssh/sky-key' ubuntu@15.165.9.140 -p 22 -o StrictHostKeyChecking=no -o ConnectTimeout=10s -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 uptime
D 01-20 09:01:25 provisioner.py:324] Retrying in 1 second...
D 01-20 09:08:40 provisioner.py:369] SSH Conection ready for 'dbg'

I 01-20 09:08:40 instance_setup.py:90] --------------------Start: internal_file_mounts --------------------
I 01-20 09:19:34 instance_setup.py:94] --------------------End:   internal_file_mounts --------------------

I 01-20 09:19:34 instance_setup.py:90] --------------------Start: setup_runtime_on_cluster --------------------
I 01-20 09:28:41 instance_setup.py:94] --------------------End:   setup_runtime_on_cluster --------------------

...
I 01-20 09:29:26 provisioner.py:473] [32mSuccessfully provisioned cluster: dbg[0m
```
Breakdown
* Provision on cloud: 91.6s
* Wait for SSH: 7min 28s
* Internal file mounts: ~11min
* setup_runtime_on_cluster: ~9min
* Remaining stages: ~1min

Total: ~30min.

Probably some of the stages took a long time due to reduced parallelism. Does supporting ~100-parallel hit the too many open files error?

**Comment by Michaelvll** [comment]
> With the latest commit, 100 nodes successfully launched!
> 
> ```
> D 01-20 09:01:08 provisioner.py:124] Provisioning 'dbg' took 91.64 seconds.
> 
> D 01-20 09:01:12 provisioner.py:366] Waiting for SSH to be available for 'dbg' ...
> D 01-20 09:01:25 provisioner.py:268] Waiting for SSH to 15.165.9.140. Try: ssh -T -i '~/.ssh/sky-key' ubuntu@15.165.9.140 -p 22 -o StrictHostKeyChecking=no -o ConnectTimeout=10s -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 uptime
> D 01-20 09:01:25 provisioner.py:324] Retrying in 1 second...
> D 01-20 09:08:40 provisioner.py:369] SSH Conection ready for 'dbg'
> 
> I 01-20 09:08:40 instance_setup.py:90] --------------------Start: internal_file_mounts --------------------
> I 01-20 09:19:34 instance_setup.py:94] --------------------End:   internal_file_mounts --------------------
> 
> I 01-20 09:19:34 instance_setup.py:90] --------------------Start: setup_runtime_on_cluster --------------------
> I 01-20 09:28:41 instance_setup.py:94] --------------------End:   setup_runtime_on_cluster --------------------
> 
> ...
> I 01-20 09:29:26 provisioner.py:473] �[32mSuccessfully provisioned cluster: dbg�[0m
> ```
> 
> Breakdown
> 
> * Provision on cloud: 91.6s
> * Wait for SSH: 7min 28s
> * Internal file mounts: ~11min
> * setup_runtime_on_cluster: ~9min
> * Remaining stages: ~1min
> 
> Total: ~30min.
> 
> Probably some of the stages took a long time due to reduced parallelism. Does supporting ~100-parallel hit the too many open files error?

Yes, it seems the too many parallelism for file mounts can open too many files, as each rsync call will open the source files and 100 parallel causes 100x opened files.

File: sky/provision/provisioner.py
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/provision/provisioner.py:1]: It is different, the new version will close the socket before we trying the ssh, while the former one will keep the socket open. This is to reduce the resources usage.

File: sky/utils/subprocess_utils.py [hunk 1]
```diff
@@ -41,6 +42,14 @@ def run_no_outputs(cmd, **kwargs):
                **kwargs)
 
 
+def get_cpu_count() -> int:
+    """Returns the number of idle CPUs."""
+    cpu_count = os.cpu_count()
+    if cpu_count is None:
+        cpu_count = 1
+    return max(4, cpu_count - 1)
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/utils/subprocess_utils.py:50]: It is to set a lower bound of the number of threads to use to 4, not the upper bound. I am renaming this function to `get_parallel_threads` instead to make it clear.

File: sky/provision/provisioner.py
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/provision/provisioner.py:1]: Got it! Sounds good.

File: sky/utils/subprocess_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/utils/subprocess_utils.py:50]: ic. Sounds good

**Review by cblmemo — APPROVED** [review]

**Review by concretevitamin — APPROVED** [review]
LGtM, thanks @Michaelvll! Does the 100 node launch finish in reasonable time now?

**Comment by Michaelvll** [comment]
> LGtM, thanks @Michaelvll! Does the 100 node launch finish in reasonable time now?

Yes, I tried it locally on my own laptop with 16 logical cores, and it finishes within **18 mins** for launching a cluster with 100 nodes and printing `hi` on aws without having my laptop hang: `sky launch -c test-100-nodes --cloud aws --cpus 2 --num-nodes 100 echo hi --use-spot --down`.

**Pull request merged**